pesto.1
.\" SPDX-License-Identifier: GPL-2.0-or-later
.\" Copyright Red Hat
.\" Author: David Gibson <david@gibson.dropbear.id.au>
.TH pesto 1

.SH NAME
.B pesto
\- Configure a running \fBpasst\fR(1) or \fBpasta\fR(1) instance.

.SH SYNOPSIS
.B pesto
[\fIOPTION\fR]... \fIPATH\fR

.SH DESCRIPTION

.B pesto
is a client to view and update the port forwarding configuration of a running
\fBpasst\fR(1) or \fBpasta\fR(1) instance.

\fIPATH\fR gives the path to the UNIX domain socket created by \fBpasst\fR or
\fBpasta\fR.  It should match the \fB-c\fR command line option given to that
instance.

.SH OPTIONS

.TP
.BR \-d ", " \-\-debug
Be verbose.

.TP
.BR \-h ", " \-\-help
Display a help message and exit.

.TP
.BR \-s ", " \-\-show
Show the forwarding configuration before and after changes are applied.

.TP
.BR \-A ", " \-\-add
Add the port forwarding specifiers following this option to the current
forwarding table, rather than replacing it.

This option can be given multiple times, as it might follow previous deletions
(see \fB--delete\fR below), and implies that all the specifiers following it,
before a further \fB--delete\fR option occurs, will be handled as additions.

See the section \fBAdding, deleting, clearing rules\fR in the \fBNOTES\fR for
more details.

.TP
.BR \-D ", " \-\-delete
Delete the port forwarding specifiers following this option from the current
forwarding table, rather than adding them to it.

This option can be given multiple times, as it might follow previous additions
(see \fB--add\fR above), and implies that all the specifiers following it,
before a further \fB--add\fR option occurs, will be handled as deletions.

See the section \fBAdding, deleting, clearing rules\fR in the \fBNOTES\fR for
more details.

.TP
.BR \-C ", " \-\-clear " " \fIpif
Clear the forwarding table associated to a given \fIpif\fR, that is, a
conceptual type of interface in \fBpasst\fR(1) or \fBpasta\fR(1) representing a
specific data path and direction.

The available \fIpif\fR names can be obtained by querying the current forwarding
configuration, which can be done by calling \fBpesto\fR(1) without options.

See the section \fBAdding, deleting, clearing rules\fR in the \fBNOTES\fR for
more details.

.TP
.BR \-t ", " \-\-tcp-ports " " \fIspec
Configure TCP port forwarding to guest or namespace. \fIspec\fR can be one of:
.RS

.TP
.BR none
Don't forward any ports

.TP
[\fIaddress\fR[\fB%\fR\fIinterface\fR]\fB/\fR]\fIports\fR ...
Specific ports to forward.  Optionally, a specific listening address
and interface name (since Linux 5.7) can be specified.  \fIports\fR
may be either:
.RS
.TP
\fBall\fR
Forward all unbound, non-ephemeral ports, as permitted by current capabilities.
No failures are reported for unavailable ports, unless no ports could be
forwarded at all.
.RE

.RS
or a comma-separated list of entries which may be any of:
.TP
\fIfirst\fR[\fB-\fR\fIlast\fR][\fB:\fR\fItofirst\fR[\fB-\fR\fItolast\fR]]
Include range. Forward port numbers between \fIfirst\fR and \fIlast\fR
(inclusive) to ports between \fItofirst\fR and \fItolast\fR.  If
\fItofirst\fR and \fItolast\fR are omitted, assume the same as
\fIfirst\fR and \fIlast\fR.  If \fIlast\fR is omitted, assume the same
as \fIfirst\fR.

.TP
\fB~\fR\fIfirst\fR[\fB-\fR\fIlast\fR]
Exclude range.  Don't forward port numbers between \fIfirst\fR and
\fIlast\fR.  This takes precedences over include ranges.

.TP
.BR auto
\fBpasta\fR only.  Only forward ports in the specified set if the
target ports are bound in the namespace. The list of ports is
periodically derived (every second) from listening sockets reported by
\fI/proc/net/tcp\fR and \fI/proc/net/tcp6\fR, see \fBproc\fR(5).
.RE

Specifying excluded ranges only implies that all other non-ephemeral
ports are forwarded. Specifying no ranges at all implies forwarding
all non-ephemeral ports permitted by current capabilities.  In this
case, no failures are reported for unavailable ports, unless no ports
could be forwarded at all.

Examples:
.RS
.TP
-t all
Forward all unbound, non-ephemeral ports as permitted by current
capabilities to the corresponding port on the guest or namespace
.TP
-t ::1/all
For the local address ::1, forward all unbound, non-ephemeral ports as
permitted by current capabilities
.TP
-t 22
Forward local port 22 to port 22 on the guest or namespace
.TP
-t 22:23
Forward local port 22 to port 23 on the guest or namespace
.TP
-t 22,25
Forward local ports 22 and 25 to ports 22 and 25 on the guest or namespace
.TP
-t 22-80
Forward local ports between 22 and 80 to corresponding ports on the guest or
namespace
.TP
-t 22-80:32-90
Forward local ports between 22 and 80 to ports between 32 and 90 on the guest or
namespace
.TP
-t 192.0.2.1/22
Forward local port 22, bound to 192.0.2.1, to port 22 on the guest or namespace
.TP
-t 192.0.2.1%eth0/22
Forward local port 22, bound to 192.0.2.1 and interface eth0, to port 22
.TP
-t %eth0/22
Forward local port 22, bound to any address on interface eth0, to port 22
.TP
-t 2000-5000,~3000-3010
Forward local ports between 2000 and 5000, except for those between 3000 and
3010
.TP
-t 192.0.2.1/20-30,~25
For the local address 192.0.2.1, forward ports between 20 and 24 and between 26
and 30
.TP
-t ~20000-20010
Forward all ports to the guest, except for the range from 20000 to 20010
.TP
-t auto
Automatically forward any ports which are bound in the namespace
.TP
-t ::1/auto
Automatically forward any ports which are bound in the namespace,
listening only on local address ::1
.TP
-t 8000-8010,auto
Forward ports in the range 8000-8010 if and only if they are bound in
the namespace
.RE
.RE

.TP
.BR \-u ", " \-\-udp-ports " " \fIspec
Configure UDP port forwarding to guest. \fIspec\fR is as described for TCP
above.

.TP
.BR \-T ", " \-\-tcp-ns " " \fIspec
Configure TCP port forwarding from target namespace to init namespace.
\fIspec\fR is as described above.

.TP
.BR \-U ", " \-\-udp-ns " " \fIspec
Configure UDP port forwarding from target namespace to init namespace.
\fIspec\fR is as described above.

.TP
.BR \-\-version
Show version and exit.

.SH NOTES

.SS Adding, deleting, clearing rules

The options \fB--add\fR, \fB--delete\fR, and \fB--clear\fR are handled as
sequential commands to manipulate the current forwarding tables. If none of them
is given, forwarding specifiers for a given table are intended as replacement of
the corresponding table. That is:

.nf
	pesto -t 1024 -U 1025
.fi

will \fBreplace\fR the current TCP inbound port forwarding table with a single
rule, forwarding port 1024, and will similarly replace the UDP outbound
forwarding table with a single forwarding rule for port 1025. This usage is a
short-hand form for:

.nf
	pesto -C HOST -t 1024 -C SPLICE -U 1025
.fi

The options \fB--add\fR and \fB--delete\fR are used to \fBadd new specific
rules or delete existing ones\fR, instead of replacing tables. For example:

.nf
	pesto -A -t 2000 -D -t 3000 -U 5000
.fi

will add a forwarding rule for inbound TCP port 2000, and delete inbound TCP
port 3000 as well as outbound UDP port 5000 from the existing set of rules.

All these options are interpreted as sequential commands and can be arbitrarily
combined. For example:

.nf
	pesto -A -t 2000 -C HOST -A -T 3000 -t 2001 -D -u 5000
.fi

will, in order:

.RS
- add inbound TCP port 2000
- clear inbound ports, reverting the addition above
- add outbound TCP port 3000
- add inbound TCP port 2001
- delete inbound UDP port 5000
.RE

.SH AUTHORS

Stefano Brivio <sbrivio@redhat.com>,
David Gibson <david@gibson.dropbear.id.au>.

.SH REPORTING BUGS

Please report issues on the bug tracker at https://bugs.passt.top/, or send a
message to the passt-user@passt.top mailing list, see https://lists.passt.top/.

.SH COPYRIGHT

Copyright Red Hat

\fBpesto\fR is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at
your option) any later version.

.SH SEE ALSO

\fBpasst\fR(1), \fBpasta\fR(1), \fBunix\fR(7).


pesto.c
// SPDX-License-Identifier: GPL-2.0-or-later

/* PESTO - Programmable Extensible Socket Translation Orchestrator
 *  front-end for passt(1) and pasta(1) forwarding configuration
 *
 * pesto.c - Main program (it's not actually extensible)
 *
 * Copyright (c) 2026 Red Hat GmbH
 * Author: Stefano Brivio <sbrivio@redhat.com>
 */

#include <arpa/inet.h>
#include <sys/prctl.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <errno.h>
#include <getopt.h>
#include <inttypes.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <unistd.h>

#include <linux/audit.h>
#include <linux/capability.h>
#include <linux/filter.h>
#include <linux/seccomp.h>

#include "common.h"
#include "seccomp_pesto.h"
#include "serialise.h"
#include "fwd_rule.h"
#include "pesto.h"
#include "log.h"

bool debug_flag = false;

static char stdout_buf[BUFSIZ];

/**
 * usage() - Print usage, exit with given status code
 * @name:	Executable name
 * @f:		Stream to print usage info to
 * @status:	Status code for exit(2)
 *
 * #syscalls:pesto exit_group fstat write
 */
static void usage(const char *name, FILE *f, int status)
{
	FPRINTF(f, "Usage: %s [OPTION]... PATH\n", name);
	FPRINTF(f,
		"\n"
		"  -A, --add		Add following specifiers to forwards\n"
		"  -D, --delete		Delete following specifiers instead\n"
		"  -C, --clear PIF	Clear forwarding table for given PIF\n"
		"  -t, --tcp-ports SPEC	TCP inbound port forwarding\n"
		"    can be specified multiple times\n"
		"    SPEC can be:\n"
		"      'none': don't forward any ports\n"
		"      [ADDR[%%IFACE]/]PORTS: forward specific ports\n"
		"        PORTS is either 'all' (forward all unbound, non-ephemeral\n"
		"        ports), or a comma-separated list of ports, optionally\n"
		"        ranged with '-' and optional target ports after ':'.\n"
		"        Ranges can be reduced by excluding ports or ranges\n"
		"        prefixed by '~'.\n"
		"        The 'auto' keyword may be given to only forward\n"
		"        ports which are bound in the target namespace\n"
		"        Examples:\n"
		"        -t all         Forward all ports\n"
		"        -t 127.0.0.1/all Forward all ports from local address\n"
		"                         127.0.0.1\n"
		"        -t 22		Forward local port 22 to 22\n"
		"        -t 22:23	Forward local port 22 to 23\n"
		"        -t 22,25	Forward ports 22, 25 to ports 22, 25\n"
		"        -t 22-80  	Forward ports 22 to 80\n"
		"        -t 22-80:32-90	Forward ports 22 to 80 to\n"
		"			corresponding port numbers plus 10\n"
		"        -t 192.0.2.1/5	Bind port 5 of 192.0.2.1\n"
		"        -t 5-25,~10-20	Forward ports 5 to 9, and 21 to 25\n"
		"        -t ~25		Forward all ports except for 25\n"
		"        -t auto	Forward all ports bound in namespace\n"
		"        -t 192.0.2.2/auto Forward ports from 192.0.2.2 if\n"
		"                          they are bound in the namespace\n"
		"        -t 8000-8010,auto Forward ports 8000-8010 if they\n"
		"                          are bound in the namespace\n"
		"  -u, --udp-ports SPEC	UDP inbound port forwarding\n"
		"    SPEC is as described for TCP above\n"
		"  -T, --tcp-ns SPEC	TCP port forwarding to init namespace\n"
		"    SPEC is as described above\n"
		"  -U, --udp-ns SPEC	UDP port forwarding to init namespace\n"
		"    SPEC is as described above\n"
		"  -s, --show		Show configuration before and after\n"
		"  -d, --debug		Print debugging messages\n"
		"  -h, --help		Display this help message and exit\n"
		"  --version		Show version and exit\n");
	exit(status);
}

/* Maximum number of pifs with rule tables */
#define MAX_PIFS	3

struct pif_configuration {
	uint8_t pif;
	char name[PIF_NAME_SIZE];
	struct fwd_table fwd;
};

struct configuration {
	uint32_t npifs;
	struct pif_configuration pif[MAX_PIFS];
};

/**
 * pif_conf_by_num() - Find a pif's configuration by pif id
 * @conf:	Configuration description
 * @pif:	pif id
 *
 * Return: pointer to the pif_configuration for @pif, or NULL if not found
 */
static struct pif_configuration *pif_conf_by_num(struct configuration *conf,
						 uint8_t pif)
{
	unsigned i;

	for (i = 0; i < conf->npifs; i++) {
		if (conf->pif[i].pif == pif)
			return &conf->pif[i];
	}

	return NULL;
}

/**
 * pif_conf_by_name() - Find a pif's configuration by name
 * @conf:	Configuration description
 * @name:	Interface name
 *
 * Return: pif_configuration for pif named @name, or NULL if not found
 */
static struct pif_configuration *pif_conf_by_name(struct configuration *conf,
						  const char *name)
{
	unsigned i;

	for (i = 0; i < conf->npifs; i++) {
		if (strcmp(conf->pif[i].name, name) == 0)
			return &conf->pif[i];
	}

	return NULL;
}

/**
 * pesto_read_rules() - Read rulestate from passt/pasta
 * @fd:		Control socket
 * @conf:	Configuration description to update
 */
static bool read_pif_conf(int fd, struct configuration *conf)
{
	struct pif_configuration *pc;
	struct pesto_pif_info info;
	uint8_t pif;
	unsigned i;

	if (read_u8(fd, &pif) < 0)
		die("Error reading from control socket");

	if (pif == PIF_NONE)
		return false;

	debug("Receiving config for PIF %"PRIu8, pif);

	if (conf->npifs >= ARRAY_SIZE(conf->pif)) {
		die("passt has more pifs than pesto can manage (max %d)",
		    ARRAY_SIZE(conf->pif));
	}

	pc = &conf->pif[conf->npifs];
	pc->pif = pif;

	if (read_all_buf(fd, &info, sizeof(info)) < 0)
		die("Error reading from control socket");

	if (info.name[sizeof(info.name)-1])
		die("Interface name was not NULL terminated");
	/* Redundant, to make static checkers happy */
	info.name[sizeof(info.name) - 1] = '\0';

	static_assert(sizeof(info.name) == sizeof(pc->name),
		      "Mismatching pif name lengths");
	memcpy(pc->name, info.name, sizeof(pc->name));
	pc->fwd.caps = ntohl(info.caps);

	pc->fwd.count = ntohl(info.count);
	if (pc->fwd.count > MAX_FWD_RULES)
		die("Too many forwarding rules");

	debug("PIF %"PRIu8": %s, %"PRIu32" rules, capabilities 0x%"PRIx32
	      ":%s%s%s%s%s%s", pc->pif, pc->name, pc->fwd.count, pc->fwd.caps,
	      pc->fwd.caps & FWD_CAP_IPV4 ? " IPv4" : "",
	      pc->fwd.caps & FWD_CAP_IPV6 ? " IPv6" : "",
	      pc->fwd.caps & FWD_CAP_TCP ? " TCP" : "",
	      pc->fwd.caps & FWD_CAP_UDP ? " UDP" : "",
	      pc->fwd.caps & FWD_CAP_SCAN ? " scan" : "",
	      pc->fwd.caps & FWD_CAP_IFNAME ? " ifname" : "");

	/* O(n^2), but n is bounded by MAX_PIFS */
	if (pif_conf_by_num(conf, pc->pif))
		die("Received duplicate interface identifier");

	/* O(n^2), but n is bounded by MAX_PIFS */
	if (pif_conf_by_name(conf, pc->name))
		die("Received duplicate interface name");

	/* NOTE: We read the fwd rules directly into fwd.rules, rather than
	 * using fwd_rule_add().  This means we can read and display rules even
	 * if something has gone wrong (in pesto or passt) and we get rules that
	 * fwd_rule_add() would reject.  It does have the side effect that we
	 * never assign socket space for the fwd rules, but we don't need that
	 * within pesto.
	 */
	for (i = 0; i < pc->fwd.count; i++) {
		if (fwd_rule_read(fd, &pc->fwd.rules[i]) < 0)
			die("Error reading from control socket");
	}

	conf->npifs++;
	return true;
}

/**
 * send_conf() - Send updated configuration to passt/pasta
 * @fd:		Control socket
 * @conf:	Updated configuration
 */
static void send_conf(int fd, const struct configuration *conf)
{
	unsigned i;

	for (i = 0; i < conf->npifs; i++) {
		const struct pif_configuration *pc = &conf->pif[i];
		unsigned j;

		if (write_u8(fd, pc->pif) < 0)
			goto fail;

		if (write_u32(fd, pc->fwd.count) < 0)
			goto fail;

		for (j = 0; j < pc->fwd.count; j++) {
			if (fwd_rule_write(fd, &pc->fwd.rules[j]) < 0)
				goto fail;
		}
	}

	if (write_u8(fd, PIF_NONE) < 0)
		goto fail;
	return;

fail:
	die_perror("Error writing to control socket");
}

/**
 * show_conf() - Show current configuration obtained from passt/pasta
 * @conf:	Configuration description
 */
static void show_conf(const struct configuration *conf)
{
	unsigned i;

	for (i = 0; i < conf->npifs; i++) {
		const struct pif_configuration *pc = &conf->pif[i];
		printf("  %s\n", pc->name);
		fwd_rules_dump(printf, pc->fwd.rules, pc->fwd.count,
			       "    ", "\n");
	}
	/* Flush stdout, so this doesn't get misordered with later debug()s */
	(void)fflush(stdout);
}

/**
 * main() - Dynamic reconfiguration client main program
 * @argc:	Argument count
 * @argv:	Arguments: socket path, operation, port specifiers
 *
 * Return: 0 on success, won't return on failure
 *
 * #syscalls:pesto socket s390x:socketcall i686:socketcall
 * #syscalls:pesto connect shutdown close
 * #syscalls:pesto exit_group fstat read write openat
 */
int main(int argc, char **argv)
{
	const struct option options[] = {
		{"debug",	no_argument,		NULL,		'd' },
		{"help",	no_argument,		NULL,		'h' },
		{"version",	no_argument,		NULL,		1 },
		{"add",		no_argument,		NULL,		'A' },
		{"delete",	no_argument,		NULL,		'D' },
		{"clear",	required_argument,	NULL,		'C' },
		{"tcp-ports",	required_argument,	NULL,		't' },
		{"udp-ports",	required_argument,	NULL,		'u' },
		{"tcp-ns",	required_argument,	NULL,		'T' },
		{"udp-ns",	required_argument,	NULL,		'U' },
		{"show",	no_argument,		NULL,		's' },
		{ 0 },
	};
	enum { MODE_CLEAR, MODE_ADD, MODE_DEL } mode = MODE_CLEAR;
	bool inbound_cleared = false, outbound_cleared = false;
	struct pif_configuration *inbound, *outbound;
	const char *optstring = "dhADC:t:u:T:U:s";
	struct sockaddr_un a = { AF_UNIX, "" };
	struct configuration conf = { 0 };
	bool update = false, show = false;
	struct pesto_hello hello;
	struct sock_fprog prog;
	int optname, ret, s;
	uint32_t s_version;

	prctl(PR_SET_DUMPABLE, 0);

	prog.len = (unsigned short)sizeof(filter_pesto) /
				   sizeof(filter_pesto[0]);
	prog.filter = filter_pesto;
	if (prctl(PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0) ||
	    prctl(PR_SET_SECCOMP, SECCOMP_MODE_FILTER, &prog))
		die("Failed to apply seccomp filter");

	/* Explicitly set stdout buffer, otherwise printf() might allocate,
	 * breaking our seccomp profile.
	 */
	if (setvbuf(stdout, stdout_buf, _IOFBF, sizeof(stdout_buf)))
		die_perror("Failed to set stdout buffer");

	fwd_probe_ephemeral();

	do {
		optname = getopt_long(argc, argv, optstring, options, NULL);

		switch (optname) {
		case -1:
		case 0:
			break;
		case 'C':
		case 't':
		case 'u':
		case 'T':
		case 'U':
		case 'A':
		case 'D':
			/* Parse these options after we've read state from
			 * passt/pasta
			 */
			update = true;
			break;
		case 's':
			show = true;
			break;
		case 'h':
			usage(argv[0], stdout, EXIT_SUCCESS);
			break;
		case 'd':
			debug_flag = true;
			break;
		case 1:
			FPRINTF(stdout, "pesto ");
			FPRINTF(stdout, VERSION_BLOB);
			exit(EXIT_SUCCESS);
		default:
			usage(argv[0], stderr, EXIT_FAILURE);
		}
	} while (optname != -1);

	if (argc - optind != 1)
		usage(argv[0], stderr, EXIT_FAILURE);

	debug("debug_flag=%d, path=\"%s\"", debug_flag, argv[optind]);

	if ((s = socket(AF_UNIX, SOCK_STREAM, 0)) < 0)
		die_perror("Failed to create AF_UNIX socket");

	ret = snprintf(a.sun_path, sizeof(a.sun_path), "%s", argv[optind]);
	if (ret <= 0 || ret >= (int)sizeof(a.sun_path))
		die("Invalid socket path \"%s\"", argv[optind]);

	ret = connect(s, (struct sockaddr *)&a, sizeof(a));
	if (ret < 0) {
		die_perror("Failed to connect to %s", a.sun_path);
	}

	debug("Connected to passt/pasta control socket");

	ret = read_all_buf(s, &hello, sizeof(hello));
	if (ret < 0)
		die_perror("Couldn't read server greeting");

	if (memcmp(hello.magic, PESTO_SERVER_MAGIC, sizeof(hello.magic)))
		die("Bad magic number from server");

	s_version = ntohl(hello.version);

	if (s_version > PESTO_PROTOCOL_VERSION) {
		die("Unknown server protocol version %"PRIu32" > %"PRIu32,
		    s_version, PESTO_PROTOCOL_VERSION);
	}

	if (!s_version) {
		if (PESTO_PROTOCOL_VERSION)
			die("Unsupported experimental server protocol");
		FPRINTF(stderr,
"Warning: Using experimental protocol version, client and server must match\n");
	}

	if (ntohl(hello.pif_name_size) != PIF_NAME_SIZE) {
		die("Server has unexpected pif name size (%"
		    PRIu32" not %"PRIu32 ")",
		    ntohl(hello.pif_name_size), PIF_NAME_SIZE);
	}

	if (ntohl(hello.ifnamsiz) != IFNAMSIZ) {
		die("Server has unexpected IFNAMSIZ (%"
		    PRIu32" not %"PRIu32 ")",
		    ntohl(hello.ifnamsiz), IFNAMSIZ);
	}

	while (read_pif_conf(s, &conf))
		;

	if (!update) {
		printf("passt/pasta configuration (%s)\n", a.sun_path);
		show_conf(&conf);
		goto noupdate;
	}

	if (show) {
		printf("Previous configuration (%s)\n", a.sun_path);
		show_conf(&conf);
	}

	inbound = pif_conf_by_name(&conf, "HOST");
	outbound = pif_conf_by_name(&conf, "SPLICE");

	optind = 0;
	do {
		struct pif_configuration *pif_to_clear;

		optname = getopt_long(argc, argv, optstring, options, NULL);

		switch (optname) {
		case 'A':
			mode = MODE_ADD;
			break;
		case 'D':
			mode = MODE_DEL;
			break;
		case 'C':
			if (!(pif_to_clear = pif_conf_by_name(&conf, optarg)))
				die("Unsupported pif name %s", optarg);

			fwd_rule_clear(&pif_to_clear->fwd);

			if (!strcmp(optarg, "HOST"))
				inbound_cleared = true;
			else if (!strcmp(optarg, "SPLICE"))
				outbound_cleared = true;

			break;
		case 't':
		case 'u':
			if (!inbound) {
				die("Can't use -%c, no inbound interface",
				    optname);
			}

			if (mode == MODE_CLEAR && !inbound_cleared) {
				fwd_rule_clear(&inbound->fwd);
				inbound_cleared = true;
			}

			fwd_rule_parse(optname, mode == MODE_DEL, optarg,
				       &inbound->fwd);
			break;
		case 'T':
		case 'U':
			if (!outbound) {
				die("Can't use -%c, no outbound interface",
				    optname);
			}

			if (mode == MODE_CLEAR && !outbound_cleared) {
				fwd_rule_clear(&outbound->fwd);
				outbound_cleared = true;
			}

			fwd_rule_parse(optname, mode == MODE_DEL, optarg,
				       &outbound->fwd);
			break;
		default:
			continue;
		}
	} while (optname != -1);

	if (show) {
		printf("Updated configuration (%s)\n", a.sun_path);
		show_conf(&conf);
	}

	send_conf(s, &conf);

noupdate:
	if (shutdown(s, SHUT_RDWR) < 0 || close(s) < 0)
		die_perror("Error shutting down control socket");

	exit(0);
}
